[LLVM][Windows] Elide `PrettyStackTrace` output for usage errors

bd1976bris opened the pull request:
On Windows, LLVM’s `reportFatalUsageError`, introduced recently (#138251), emits bug report prompt due to the `PrettyStackTrace` signal handler, initialized via `InitLLVM`. This occurs when `sys::RunInterruptHandlers()` is called from `reportFatalUsageError`.

This behavior is misleading for usage errors. For example, one of Sony’s customers filed a bug after specifying an invalid LTO cache directory - a clear usage error - because the toolchain output included instructions to report a bug.

This patch suppresses `PrettyStackTrace` output for usage errors by adding a flag to `sys::RunInterruptHandlers()` to indicate whether signal handlers should be executed.

LLVM Issue: https://github.com/llvm/llvm-project/issues/140953
Internal Tracker: TOOLCHAIN-17744

tru reviewed (approved):
LGTM. This has bitten us before as well.

nikic commented on llvm/include/llvm/Support/ErrorHandling.h:
```diff
@@ -61,11 +61,14 @@ struct ScopedFatalErrorHandler {
 [[noreturn]] LLVM_ABI void report_fatal_error(const Twine &reason,
-                                              bool gen_crash_diag = true);
+                                              bool gen_crash_diag = true,
+                                              bool is_usage_error = false);
```
These changes don't make sense: gen_crash_diag *is* already the flag that indicates whether this is a usage error and controls whether a stack trace is displayed.

Calling reportFatalUsageError() will correctly not display a stack trace and not ask for a bug report on Linux. If this is not the case on Windows, you need to identify why there is a difference in behavior. Adding an extra flag here is certainly the wrong thing to do though.

bd1976bris replied:
Thanks. I thought that adding a new flag would mean that the behaviours of 1. `report_fatal_error` calling `abort()` and 2. The signal handler having usage or internal error behaviour were independent - which might be required. I'll update to just use gen_crash_diag - that will make for a smaller patch anyway.

As to why Windows has different behaviour - on Linux LLVM once called the signal handler for `report_fatal_error` IIUC but this seems to have changed in: https://github.com/llvm/llvm-project/commit/288999b82959d6a057bfd478a196961215baf792.

bd1976bris replied:
@nikic I'm just using `gen_crash_diag` now. Any good?

nikic replied:
> As to why Windows has different behaviour - on Linux LLVM once called the signal handler for `report_fatal_error` IIUC but this seems to have changed in: [288999b](https://github.com/llvm/llvm-project/commit/288999b82959d6a057bfd478a196961215baf792).

So, can we do the same change for Windows? Basically pass false to Cleanup in the Windows RunInterruptHandlers, or something along those lines?

bd1976bris replied:
@nikic I have just realised that I might be able to be a bit braver and not pass anything into the signal handlers but instead just not call them in Windows `signal.inc`. Let me try that...

bd1976bris replied:
> > As to why Windows has different behaviour - on Linux LLVM once called the signal handler for `report_fatal_error` IIUC but this seems to have changed in: [288999b](https://github.com/llvm/llvm-project/commit/288999b82959d6a057bfd478a196961215baf792).
> 
> So, can we do the same change for Windows? Basically pass false to Cleanup in the Windows RunInterruptHandlers, or something along those lines?

Seems to work and much better. Thanks!

nikic replied:
My remaining question here would be whether we actually need the ExecuteSignalHandlers argument on RunInterruptHandlers, or can always assume it is false like on Linux. If I understood things correctly, the way this is supposed to work is that if GenCrashDiag is true then the following abort() is going to run the signal handlers anyway?

bd1976bris replied:
Good questions. 

>If I understood things correctly, the way this is supposed to work is that if GenCrashDiag is true then the following abort() is going to run the signal handlers anyway?

Yes. To keep the previous behaviour where the signal handlers are always run. I need to poke at it a bit. Perhaps the output is OK on Windows without running the signal handlers. I don't understand why the difference exists between the two platforms.

bd1976bris commented:
I wonder if we need a release note for this behaviour change? I've added on in case.

arsenm commented on lld/test/COFF/lto-cache-errors.ll:
```diff
@@ -1,15 +1,26 @@
-; REQUIRES: x86, non-root-user
-;; Not supported on windows since we use permissions to deny the creation
-; UNSUPPORTED: system-windows
+;; The output is OS-specific, so this test is limited to Windows.
```
Switching a test from "not windows" to "only windows" is suspicious. Is it just the error message phrasing? Can you follow the pattern of other tests that depend on error strings, e.g. llvm/test/tools/llvm-ar/error-opening-permission.test

bd1976bris replied:
Thanks. I made the test Windows specific because there is additional output on Windows (that I'm trying to eliminate in this patch) so that seems a better platform to run this test on than Linux. It's stell testing the same failure condition. I'll try to either make the test work on both platforms or add a separate Windows specific test.

bd1976bris replied:
I have had to change the test strategy for this to use the invalid pass pipeline errors, as they occur in a single-threaded context*. The test for that in LLD works on both platforms.

* #140955 was reverted. Digging into why the test failed, I suspect that reportFatalUsageError can cause a crash if threads call reportFatalUsageError simultaneously (the ThinLTO cache dir is used from a multi-threaded context). I have observed this on Windows. It may also apply on Linux.

arsenm commented on llvm/include/llvm/Support/Error.h:
```diff
@@ -744,7 +744,7 @@ template <class T> class [[nodiscard]] Expected {
 /// instead.
 [[noreturn]] LLVM_ABI void report_fatal_error(Error Err,
-                                              bool gen_crash_diag = true);
+                                              bool GenCrashDiag = true);
```
Unrelated rename

bd1976bris replied:
Thanks. I'll remove anything that is just a tidy-up and not required.

bd1976bris replied:
Done.

bd1976bris committed:
[LLVM][Windows] Elide `PrettyStackTrace` output for usage errors

On Windows, LLVM’s `reportFatalUsageError` (#138251) may emit a
bug report prompt due to the `PrettyStackTrace` signal handler,
initialized via `InitLLVM`. This occurs when `RunInterruptHandlers()`
is called from `reportFatalUsageError`.

This behavior is misleading for usage errors. For example, one
of Sony’s customers filed a bug after specifying an invalid LTO
cache directory - a clear usage error - because the toolchain
output included instructions to report a bug.

This patch suppresses `PrettyStackTrace` output for usage errors by
adding a flag to `sys::RunInterruptHandlers()` to indicate whether
signal handlers should be executed.

To test this, I have modified the invalid LTO pipeline errors to call
`reportFatalUsageError`, and I have updated the existing LLD test to
additionally verify that no bug report message has been emitted.

LLVM Issue: https://github.com/llvm/llvm-project/issues/140953
Internal Tracker: TOOLCHAIN-17744
